### PR TITLE
policy-compiler: mocks deployment

### PR DESCRIPTION
**Pull request by rohithdv**
This PR has updated scripts in test/services/pilot folder to enable deployment of catalog connector and policy manager mocks in custom openshift project which are useful for testing new features related to policy compiler and connectors locally before integrating the changes with the data mesh.

The scripts added here are based on the scripts in connectors/helpers/ folder.

The following sequence of commands can be followed to build and deploy updated mocks.

make dockerall (in test/services/pilot) - builds docker images for the mocks and uploads to relevant docker registry (either ghcr.io or ibmcloud registry)
Set KUBE_NAMESPACE=<OCPROJECTNAME>
make deploy (in test/services/pilot) - deploys the relevant docker registry (either ghcr.io or ibmcloud registry) to openshift project.

### CODE REVIEW

File: test/services/pilot/deploy_mocks.sh [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        kubectl create secret docker-registry cloud-registry \
+            --docker-server="$DOCKER_HOSTNAME" \
+            --docker-username="$DOCKER_USERNAME" \
+            --docker-password="$DOCKER_PASSWORD" \
+            -n $KUBE_NAMESPACE
+
+        kubectl patch serviceaccount default -p \
+            '{"imagePullSecrets": [{"name": "cloud-registry"}]}' \
+            -n $KUBE_NAMESPACE
+}
```
Comment (hunchback) [inline test/services/pilot/deploy_mocks.sh:33]: already handled by `third_party/registry`

File: test/services/pilot/deploy_mocks.sh [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        printf "\nRemoving kubectl mock resources on active cluster"
+        kubectl delete --ignore-not-found -f $ROOT_DIR/manager/config/prod/deployment_configmap.yaml 
+        #cd $ROOT_DIR/test/services/pilot/deploy
+        #kustomize build patch/$REGISTRY | kubectl delete -f - || true
+        kustomize_build delete true
+        #cd -
+}
+
+mocks_create() {
+        kubectl apply -f $ROOT_DIR/manager/config/prod/deployment_configmap.yaml
```
Comment (hunchback) [inline test/services/pilot/deploy_mocks.sh:61]: is part of the manager deployment

File: test/services/pilot/deploy_mocks.sh [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        printf "\nSleep ...\n"
+        sleep 10
+
+        printf "\nGetting current status of active cluster. Please wait...\n"
+        kubectl cluster-info
+        kubectl get all
+
+        printf "\n(TIP:) You can use the command \'watch kubectl get all\' to continuously monitor the cluster resources!\n"
+        printf "\nThe deployment script has completed successfully!\n"
+}
```
Comment (hunchback) [inline test/services/pilot/deploy_mocks.sh:78]: duplicated from the connectors deployment and should be deleted

File: test/services/pilot/deploy_mocks.sh [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+docker)
+    echo "Registry docker not supported" || exit 1
+    ;;
+localhost)
+    echo "Registry localhost not supported" || exit 1
+    ;;
+*)
+    REGISTRY=ibmcloud 
+    ;;
+esac
```
Comment (hunchback) [inline test/services/pilot/deploy_mocks.sh:90]: not relevant, just use registry defined in DOCKER_HOSTNAME

File: connectors/vault/vault_utils/vault_utils.go [hunk 5]
```diff
@@ -135,3 +137,24 @@ func (vlt *VaultConnection) GetFromVault(innerVaultPath string) (string, error)
 
 	return string(b), nil
 }
+
+// GetFromVault returns the credentials from vault as json
```
Comment (hunchback) [inline connectors/vault/vault_utils/vault_utils.go:141]: comment should start with `AddToVault`

File: test/services/pilot/Makefile [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  dockerbuild:: ; cd $1 && make docker-build
+endef
+
+define docker-push-target
+  dockerpush:: ; cd $1 && make docker-push
+endef
+
+define docker-rmi-target
+  dockerrmi:: ; cd $1 && make docker-rmi
+endef
```
Comment (hunchback) [inline test/services/pilot/Makefile:43]: why are these needed? how are these used?

File: test/services/pilot/catalog-credential-mock/go.mod [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+replace github.com/ibm/the-mesh-for-data/connectors/vault v0.0.0 => ../../../../connectors/vault
+
+go 1.13
+
+require (
+	github.com/golang/protobuf v1.4.2
+	github.com/hashicorp/vault/api v1.0.4
+	github.com/ibm/the-mesh-for-data/connectors/vault v0.0.0
+	google.golang.org/grpc v1.28.1
+)
```
Comment (hunchback) [inline test/services/pilot/catalog-credential-mock/go.mod:16]: please remove, see how @froesef builds everything on the host so as to shorted build time by using Travis cache. Once build otside there is no need to maintain a prove go.mod

File: connectors/vault/vault_utils/vault_utils.go (hunk 5 above)
Comment (etykha) [reply connectors/vault/vault_utils/vault_utils.go:141]: the content also should be changed to "adding" credentials instead of "getting" them

File: test/services/pilot/deploy_mocks.sh (hunk 2 above)
Comment (etykha) [reply test/services/pilot/deploy_mocks.sh:61]: mocks and connectors should be able to run separately from the manager in a different namespace

File: test/services/pilot/deploy_mocks.sh (hunk 3 above)
Comment (etykha) [reply test/services/pilot/deploy_mocks.sh:78]: why getting the status should not be part of all deployment processes? Can we reuse this function from connectors deployment script here?

File: test/services/pilot/deploy_mocks.sh (hunk 1 above)
Comment (rohithdv) [reply test/services/pilot/deploy_mocks.sh]: fixed. Thanks.

File: test/services/pilot/deploy_mocks.sh (hunk 3 above)
Comment (rohithdv) [reply test/services/pilot/deploy_mocks.sh:58]: Due to duplication as mentioned by @hunchback, I am removing this function for now. Thanks.

File: test/services/pilot/deploy_mocks.sh (hunk 4 above)
Comment (rohithdv) [reply test/services/pilot/deploy_mocks.sh]: fixed. Thanks.

File: connectors/vault/vault_utils/vault_utils.go (hunk 5 above)
Comment (rohithdv) [reply connectors/vault/vault_utils/vault_utils.go]: fixed. Thanks.

File: test/services/pilot/Makefile (hunk 6 above)
Comment (rohithdv) [reply test/services/pilot/Makefile:43]: These targets are used for the usual operations but with respect to the 3 mocks (representing the catalog manager, credential manager, policy manager GRPC interfaces)
make build : builds the catalog connector and policy manager mock. 
make run, terminate : executes/terminates the catalog connector and policy manager mock. 
make dockerall : build mocks docker images for catalog connector and policy manager mock. 
make clean: cleans the mocks related logs
make docker-rmi: remove docker images for catalog connector and policy manager mock. 
make docker-push: push docker images for catalog connector and policy manager mock to the relevant docker registry

Also this makefile is integrated from the pkg/policy-compiler makefile as well so that all the connectors, mocks, external servers (like opa, egeria), are executed in sequence (in the scenario of local machine testing) and we can quickly check if there are any issues with the connectors execution logic (without need to deploy on the cluster env).

File: test/services/pilot/deploy_mocks.sh [hunk 8]
```diff
@@ -47,7 +47,6 @@ mocks_create() {
 
 undeploy() {
         mocks_delete
-        kube_cluster_info
```
Comment (etykha) [inline test/services/pilot/deploy_mocks.sh]: why do you remove the cluster info after deployment? Mocks will be deployed separately from the manager. Is there another place where you ask for cluster info? if not why you prefer to remove this part of deployment?

File: test/services/pilot/catalog-credential-mock/go.mod (hunk 7 above)
Comment (rohithdv) [reply test/services/pilot/catalog-credential-mock/go.mod]: fixed by removing the private  go.mod files from catalog-crendential-mock/ folder and also from connectors/vault and connectors/opa folders. Thanks.

File: test/services/pilot/deploy_mocks.sh (hunk 8 above)
Comment (rohithdv) [reply test/services/pilot/deploy_mocks.sh]: added kube_cluster_info back to the script. Thanks.

File: connectors/vault/user_vault_init.sh [hunk 9]
```diff
@@ -22,7 +22,7 @@ source $DIR_ROOT/pkg/policy-compiler/policy-compiler.env
 
 export VAULT_BIN=$DIR_ABSTOOLBIN/vault
 #export USER_VAULT_KAFKA_ASSET_KEY=87ffdca3-8b5d-4f77-99f9-0cb1fba1f73f/01c6f0f0-9ffe-4ccc-ac07-409523755e72
-export USER_VAULT_KAFKA_ASSET_KEY="{\"catalog_id\":\"87ffdca3-8b5d-4f77-99f9-0cb1fba1f73f\",\"asset_id\":\"01c6f0f0-9ffe-4ccc-ac07-409523755e72\"}"
+export USER_VAULT_KAFKA_ASSET_KEY="{\"catalog_id\":\"fb4e5fc8-266c-4cf7-a29f-33e3c9d5d091\",\"asset_id\":\"05df45f2-5927-4534-bdff-0d4b923baeea\"}"
```
Comment (roee88) [inline connectors/vault/user_vault_init.sh:25]: Why is this keep changing? can't we fix it to something?

File: connectors/vault/vault_connector.go [hunk 10]
```diff
@@ -11,14 +11,15 @@ import (
...
 	"google.golang.org/grpc"
 )
 
-var vault VaultConnection
+var vault vltutils.VaultConnection
 
-const defaultTimeout = "180"
-const defaultPort = "50083" //synched with vault_connector.yaml
+//const defaultTimeout = "180"
+//const defaultPort = "50083" //synched with vault_connector.yaml
```
Comment (roee88) [inline connectors/vault/vault_connector.go:22]: remove commented out code

File: connectors/vault/vault_connector.go [hunk 11]
```diff
@@ -42,17 +43,17 @@ func (s *server) GetCredentialsInfo(ctx context.Context, in *pb.DatasetCredentia
 }
 
 func main() {
-	vaultAddress := getEnv(VaultAddressKey)
-	timeOutInSecs := getEnvWithDefault(VaultTimeoutKey, defaultTimeout)
-	timeOutSecs, err := strconv.Atoi(timeOutInSecs)
-	port := getEnvWithDefault(VaultConnectorPortKey, defaultPort)
+	vaultAddress := vltutils.GetEnv(vltutils.VaultAddressKey)
+	timeOutInSecs := vltutils.GetEnvWithDefault(vltutils.VaultTimeoutKey, vltutils.DefaultTimeout)
```
Comment (roee88) [inline connectors/vault/vault_connector.go:47]: Neither are really "vault" utilities. Look for a package that has similar or create such functions in `pkg` and reuse (e.g., in `pkg/envvars`).

File: connectors/vault/vault_connector.go [hunk 12]
```diff
@@ -42,17 +43,17 @@ func (s *server) GetCredentialsInfo(ctx context.Context, in *pb.DatasetCredentia
...
+	timeOutSecs, _ := strconv.Atoi(timeOutInSecs)
+	port := vltutils.GetEnvWithDefault(vltutils.VaultConnectorPortKey, vltutils.DefaultPort)
 
-	log.Printf("Vault address env variable in %s: %s\n", VaultAddressKey, vaultAddress)
-	log.Printf("VaultConnectorPort env variable in %s: %s\n", VaultConnectorPortKey, port)
+	log.Printf("Vault address env variable in %s: %s\n", vltutils.VaultAddressKey, vaultAddress)
+	log.Printf("VaultConnectorPort env variable in %s: %s\n", vltutils.VaultConnectorPortKey, port)
```
Comment (roee88) [inline connectors/vault/vault_connector.go:52]: vault connector is specific to the connector, why is it in vault utils?

File: connectors/vault/vault_utils/vault_utils.go [hunk 13]
```diff
@@ -0,0 +1,213 @@
+// Copyright 2020 IBM Corp.
```
Comment (roee88) [inline connectors/vault/vault_utils/vault_utils.go:1]: Hard to comment since I don't know what you changed, next time please `git mv`.

File: connectors/vault/vault_utils/vault_utils.go [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+	if jsonErr != nil {
+		msg := "Error marshaling credentials to json for " + vaultPath + ":" + jsonErr.Error()
+		return "", errors.New(msg)
+	}
+
+	return string(b), nil
+}
+
+// GetFromVault2 returns the credentials from vault as json given outer and inner vault path parameters
+func (vlt *VaultConnection) GetFromVault2(outerVaultPath string, innerVaultPath string) (string, error) {
```
Comment (roee88) [inline connectors/vault/vault_utils/vault_utils.go:160]: GetFromVault2?

File: connectors/vault/vault_utils/vault_utils.go [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+		Config: config,
+	}
+
+	client, _ := connection.InitVault()
+	connection.Client = client
+
+	return connection
+}
+
+func CreateVaultConnection2(vaultaddress string) VaultConnection {
```
Comment (roee88) [inline connectors/vault/vault_utils/vault_utils.go:75]: CreateVaultConnection2?

File: manager/config/prod/deployment_configmap.yaml [hunk 16]
```diff
@@ -9,24 +9,31 @@ data:
   SECRET_PROVIDER_URL: "http://secret-provider.m4d-system.svc.cluster.local:5555/get-secret"
   SECRET_PROVIDER_ROLE: "demo"
   #catalog connector
-  CATALOG_CONNECTOR_URL: "egr-connector:50084"
-  CATALOG_PROVIDER_NAME: "EGERIA"
+  # CATALOG_CONNECTOR_URL: "egr-connector:50084"
+  # CATALOG_PROVIDER_NAME: "EGERIA"
+  CATALOG_CONNECTOR_URL: "wkc-connector:50090"
+  CATALOG_PROVIDER_NAME: "WKC"
```
Comment (roee88) [inline manager/config/prod/deployment_configmap.yaml:15]: why?

File: manager/config/prod/deployment_configmap.yaml [hunk 17]
```diff
@@ -9,24 +9,31 @@ data:
...
+  MAIN_POLICY_MANAGER_NAME: "WKC"
   USE_EXTENSIONPOLICY_MANAGER: "false"
   #EXTENSIONS_POLICY_MANAGER_CONNECTOR_URL: "opa-connector:50082"
   #EXTENSIONS_POLICY_MANAGER_NAME: "OPA"
 
   #internal vault related variables
   VAULT_ADDRESS: "http://vault.m4d-system.svc.cluster.local:8200/"
   VAULT_DATASET_HOME: "m4d/dataset-creds/"
-  VAULT_USER_HOME: "m4d/user-creds/"
+  VAULT_USER_HOME: "secret/m4d/user-creds/"
```
Comment (roee88) [inline manager/config/prod/deployment_configmap.yaml:36]: why?

File: manager/config/prod/deployment_configmap.yaml [hunk 18]
```diff
@@ -42,3 +49,8 @@ data:
   #external vault  variables
   USER_VAULT_ADDRESS: "http://vault.m4d-system.svc.cluster.local:8200/"
   USER_VAULT_PATH: "external"
+  APPID: "datauser1/notebook-with-kafka"
```
Comment (roee88) [inline manager/config/prod/deployment_configmap.yaml:52]: what is it?

**Review by roee88 — CHANGES_REQUESTED** [review]
General comments:
1. @hunchback and me to discuss go.mod for connectors
1.  Remove commented out code. Git keeps history for us.
1. Try to use `git mv` when moving files around and commit after move 

Specific to this PR:
1. It does much much more than mocks deployment, I don't understand why
1. Can you check if https://github.com/IBM/the-mesh-for-data/pull/61/files#diff-ba45de15fa03d5b695decc764b19084fc8ae2b778b65ea0645739eb3a8f714a0 includes mock deployment and sync with @froesef to avoid duplications
1. Vault utilities originated from the `manager` and is now a complex fork here. Instead, create a PR to have real vault utilities in `pkg` and those should be shared by whatever component that needs them.

**Comment by rohithdv** [comment]
> General comments:
> 
> 1. @hunchback and me to discuss go.mod for connectors
> 2. Remove commented out code. Git keeps history for us.
> 3. Try to use `git mv` when moving files around and commit after move
> 
> Specific to this PR:
> 
> 1. It does much much more than mocks deployment, I don't understand why
> 2. Can you check if https://github.com/IBM/the-mesh-for-data/pull/61/files#diff-ba45de15fa03d5b695decc764b19084fc8ae2b778b65ea0645739eb3a8f714a0 includes mock deployment and sync with @froesef to avoid duplications
> 3. Vault utilities originated from the `manager` and is now a complex fork here. Instead, create a PR to have real vault utilities in `pkg` and those should be shared by whatever component that needs them.


Regarding Point 2 comment in "Specific to this PR section", what I think @froesef  is doing is related to mocks the pilot uses (to mock policy-compiler). What mocks we use are the mocks for pilot itself. These are quite different in my opinion. I think there is no duplication involved here as the components which are mocked are different.

**Comment by rohithdv** [comment]
> General comments:
> 
> 1. @hunchback and me to discuss go.mod for connectors
> 2. Remove commented out code. Git keeps history for us.
> 3. Try to use `git mv` when moving files around and commit after move
> 
> Specific to this PR:
> 
> 1. It does much much more than mocks deployment, I don't understand why
> 2. Can you check if https://github.com/IBM/the-mesh-for-data/pull/61/files#diff-ba45de15fa03d5b695decc764b19084fc8ae2b778b65ea0645739eb3a8f714a0 includes mock deployment and sync with @froesef to avoid duplications
> 3. Vault utilities originated from the `manager` and is now a complex fork here. Instead, create a PR to have real vault utilities in `pkg` and those should be shared by whatever component that needs them.

Regarding Point 3 in "Specific to this PR" section, vault utils was already there in the code earlier as per this PR created by Ety (Pls check here: https://github.ibm.com/data-mesh-research/datamesh/pull/403/files#r3669272). I have just placed it in a separate folder for convenience with some minor changes. I personally think it should be named as "vault connector" utils as it is only used  in the vault connector. It is not used anywhere else. Pls suggest how to proceed.

**Comment by rohithdv** [comment]
> General comments:
> 
> 1. @hunchback and me to discuss go.mod for connectors
> 2. Remove commented out code. Git keeps history for us.
> 3. Try to use `git mv` when moving files around and commit after move
> 
> Specific to this PR:
> 
> 1. It does much much more than mocks deployment, I don't understand why
> 2. Can you check if https://github.com/IBM/the-mesh-for-data/pull/61/files#diff-ba45de15fa03d5b695decc764b19084fc8ae2b778b65ea0645739eb3a8f714a0 includes mock deployment and sync with @froesef to avoid duplications
> 3. Vault utilities originated from the `manager` and is now a complex fork here. Instead, create a PR to have real vault utilities in `pkg` and those should be shared by whatever component that needs them.

Regarding Point 1 in "Specific to this PR:" section, since this PR is quite old and we had a demo in between, we had to make some changes which is not exactly relevant to this PR but was relevant to the demo due to the time deadline involved. Though I feel all these changes are needed, pls let me know if you want to close this PR and create another PR just focused on mocks deployment.

File: manager/config/prod/deployment_configmap.yaml (hunk 17 above)
Comment (rohithdv) [reply manager/config/prod/deployment_configmap.yaml:36]: when I tried to add to the m4d path it was not working as I think the m4d path is not enabled by default(?) when I last checked. The 'secret' path is enabled by default and hence I used that. Anyways I will check again and confirm on this point. Thanks.

File: manager/config/prod/deployment_configmap.yaml (hunk 18 above)
Comment (rohithdv) [reply manager/config/prod/deployment_configmap.yaml:52]: In a normal flow, APPID is passed to connectors from pilot (https://github.com/IBM/the-mesh-for-data/blob/master/pkg/connectors/protos/data_catalog_request.proto (2) https://github.com/IBM/the-mesh-for-data/blob/master/pkg/connectors/protos/data_credential_request.proto (3) https://github.com/IBM/the-mesh-for-data/blob/acb9a2c96963ac263a3867431b56609a02bc688b/pkg/connectors/protos/policy_manager_request.proto#L36). 

Since we are trying to mock pilot here, I needed a place where the APPID could be defined for creating a mock pilot requests. Hence I used this variable in the deployment_configmap.yaml. It could have been defined in the files related to mockup code as well (which leads to duplication of the configuration). Please let me know if I need to move it to files related to mockup code i.e. in the following locations:

(1) https://github.com/rohithdv/the-mesh-for-data/blob/baa89451b9b901ddbfe3348e72c404a0e2f8198b/test/services/pilot/catalog-credential-mock/mockup_pilot_catalog_con.go#L62 

(2) https://github.com/rohithdv/the-mesh-for-data/blob/baa89451b9b901ddbfe3348e72c404a0e2f8198b/test/services/pilot/policy-manager-mock/mockup_pilot.go#L89 ?)

File: manager/config/prod/deployment_configmap.yaml (hunk 16 above)
Comment (rohithdv) [reply manager/config/prod/deployment_configmap.yaml:15]: We used this branch for the demo on Oct 15. Hence we had to make the default connectors to point to WKC.

File: connectors/vault/vault_utils/vault_utils.go (hunk 13 above)
Comment (rohithdv) [reply connectors/vault/vault_utils/vault_utils.go:1]: Sure thanks.  I just moved vault_utils.go to vault_utils folder and then there were some minor changes done. Will use git mv next time around.

File: connectors/vault/user_vault_init.sh (hunk 9 above)
Comment (rohithdv) [reply connectors/vault/user_vault_init.sh:25]: As we use different assets and CPD instances during different times, we need to update the assetids frequently.

Comment (roee88) [reply connectors/vault/user_vault_init.sh:25]: Fine an alternative that doesn't require changing the source code frequently, such as adding a parameter to this bash script,

File: manager/config/prod/deployment_configmap.yaml (hunk 16 above)
Comment (roee88) [reply manager/config/prod/deployment_configmap.yaml:15]: We are not changing the prod config just because of a demo

File: manager/config/prod/deployment_configmap.yaml (hunk 18 above)
Comment (roee88) [reply manager/config/prod/deployment_configmap.yaml:52]: Yes, it shouldn't be part of the deployment configmap.

**Comment by roee88** [comment]
Hi @rohithdv

> Regarding Point 3 in "Specific to this PR" section, vault utils was already there in the code earlier as per this PR created by Ety. I have just placed it in a separate folder for convenience with some minor changes. I personally think it should be named as "vault connector" utils as it is only used in the vault connector. It is not used anywhere else. Pls suggest how to proceed.

It was forked from https://github.com/IBM/the-mesh-for-data/blob/master/manager/controllers/utils/vault.go. A separate PR should move common vault utilities to `pkg/vault`. I don't know how this file got in with methods like `AddToVault2` :). I would to review it separately, hence the request for a separate PR for vault utilities. You can open an issue for it so we can work this out later.

> Regarding Point 1 in "Specific to this PR:" section, since this PR is quite old and we had a demo in between, we had to make some changes which is not exactly relevant to this PR but was relevant to the demo due to the time deadline involved. Though I feel all these changes are needed, pls let me know if you want to close this PR and create another PR just focused on mocks deployment.

Please open a new PR with just mocks deployment. It doesn't matter if it's all required, as you can see there are unrelated changes that block merging this. Keeping PRs focused is the way to go! :) I think that opening an issue before creating a PR will help in the future to keep things in focus.

Closing this PR.
